### ISSUE TITLE
Tetris (Motorola 96x65) keybind issue

### ISSUE THREAD

**Issue by TwoSpacesSG** [TASEmulators/freej2me-plus#48]
The 2 key (8 on computer keyboard) both rotates the piece and soft-drops it. I assume it's something to do with game actions because the Up key doesn't have this issue.

**Comment by TwoSpacesSG** [comment]
Probably related to this, in the game called Landers, the 5 key works as a menu confirm key on Nokia, but not on FJ+.

**Comment by TwoSpacesSG** [comment]
Tetris (Motorola 96x65) is now broken as of the current build, moving the piece noticeably lags the game, with "Recursive repaint attempted" in the console. Landers continues to have the aforementioned issue.

**Comment by AShiningRay** [comment]
> Probably related to this, in the game called Landers, the 5 key works as a menu confirm key on Nokia, but not on FJ+.

Yeah for this to work, Canvas would have to map the numkeys to the same action keys they can "functionally" map to. Previously, the 2,4,6,8 num keys mapped to UP, LEFT, RIGHT, DOWN canvas keys, but num 5 mapped to KEY_NUM5... i think i had tested a game that required FIRE and NUM5 to have different game mappings, though i don't remember which. 

And now with commit f9740526c57b3e6de955da10f7f8b6ad554bec74, even those movement keys are mapped to their respective Canvas mappings because LCDUI navigation requires it (and this is as it should be), otherwise textboxes and textfields would lock the UI in place and prevent scrolling... if their inputs was made to simulate a phone's typing method, this separation would still be required anyway, but with arrow keys being used for navigation.

I'll see if duplicating those functional binds for getGameAction() only helps, though it will likely break something else... might even break tetris further, as it's already exhibiting multiple functions in a single button press.

**Comment by AShiningRay** [comment]
After some internal discussions, this is actually a game bug. Maybe due to being early MIDP where key handling wasn't as standardized and vendor VMs were less feature-complete. What happens with this game's input handling is this:

```
// GetGameAction block, for action keys, this works normally
switch(this.getGameAction(var1)) { 
      case 1: // Up
         var4 = 3; // Rotate counterclockwise
         break;
      case 2: // Left
         var2 = -1; // Move left
      case 3: // no canvas key matches against this
      case 4: // same for this
      default:
         break;
      case 5: // Right
         var2 = 1; // Move right
         break;
      case 6: // Down
         var3 = true; // Drop faster
} // End of getGameAction block
  
// This is where the issue lies, matching against the phone's numpad
switch(var1) {
      case 50: // NUM 2
      case 56: // NUM 8
         var3 = true; // Drop faster
         break;
      case 51: // NUM 3
         var4 = 3; // Rotate counterclockwise
         break;
      case 52: // NUM 4
         var2 = -1; // Move left
         break;
      case 53: // NUM 5
         var2 = 1; // Move RIGHT???
         break;
      case 54: // NUM 6
         var4 = 1; // Rotate clockwise
      case 55:
}
```

Game calls for getGameAction for abstracted inputs (which also abstracts the numpad, which the vast majority of jars expect it to do) so NUM_2 matches the action it should, which is rotating the block, but right after it matches the games weird NUM_2 case below that makes the block fall faster. The whole direct numpad handling is weird as can be seen above.

Marking it as wontfix because even with a compatibility flag the keypad handling is still strange and non-standard.
